src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -586,23 +598,27 @@ public static int LastIndexOf(ref byte searchSpace, int searchSpaceLength, ref b
                 {
                     Vector256<byte> cmpCh1 = Vector256.Equals(ch1, Vector256.LoadUnsafe(ref searchSpace, (nuint)offset));
                     Vector256<byte> cmpCh2 = Vector256.Equals(ch2, Vector256.LoadUnsafe(ref searchSpace, (nuint)(offset + ch1ch2Distance)));
-                    uint mask = (cmpCh1 & cmpCh2).AsByte().ExtractMostSignificantBits();
+                    Vector256<byte> cmpAnd = (cmpCh1 & cmpCh2).AsByte();
 
-                    while (mask != 0)
+                    if (!cmpAnd.IsAllZero())
                     {
-                        // unlike IndexOf, here we use LZCNT to process matches starting from the end
-                        int bitPos = 31 - BitOperations.LeadingZeroCount(mask);
-                        if (valueTailNLength == 1 || // we already matched two bytes
-                            SequenceEqual(
-                                ref Unsafe.Add(ref searchSpace, offset + bitPos + 1),
-                                ref valueTail,
-                                valueTailNLength))
+                        uint mask = cmpAnd.ExtractMostSignificantBits();
+                        while (mask != 0)
                         {
-                            return bitPos + offset;
-                        }
+                            // unlike IndexOf, here we use LZCNT to process matches starting from the end
+                            int bitPos = 31 - BitOperations.LeadingZeroCount(mask);
+                            if (valueTailNLength == 1 || // we already matched two bytes
+                                SequenceEqual(
+                                    ref Unsafe.Add(ref searchSpace, offset + bitPos + 1),
+                                    ref valueTail,
+                                    valueTailNLength))
+                            {
+                                return bitPos + offset;
+                            }
 
-                        // Clear the highest set bit.
-                        mask &= ~(uint)(1 << bitPos);
+                            // Clear the highest set bit.
+                            mask &= ~(uint)(1 << bitPos);
```

Comment:
This might be another where a helper is beneficial to have a helper, if we expect it to be more widespread and/or specially optimizable

Reply:
Particularly given every usage is likely to be prefixed with a comment saying `// clear the highest set bit`

Reply:
Bmi2.ZeroHighBits could be used here